### ISSUE TITLE
routerLinkActive within ngif block leads to strange behaviour

### ISSUE THREAD

**Issue by stefanholzapfel** [angular/angular#14227]
**I'm submitting a ...**  
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Following code:
```
<div *ngIf="currentUser !== null" class="...">
    <div id="notifications">
        <core-toasts></core-toasts>
    </div>
...
         <li *ngFor="let moduleTab of moduleTabs" [routerLinkActive]="['active']">
              <a [routerLink]="moduleTab.id ? moduleTab.route + '/' + moduleTab.id : moduleTab.route" [routerLinkActive]="['active']">
                      <span>...</span>
               </a>
         </li>
...
</div>
```
...leads to the effect, that all contained component's ngOnInit hooks (e.g. the core-toast's) are running multiple times, even though the currentUser is set from null to a value only once. The ngOnDestroy is NOT executed, so all components seem to stay alive.

The currentUser is a private variable in the component which is set by an RxJs Observable (coming from Angular-Meteor, patched into the zone via zone.run).

When I remove ONLY the [routerLinkActive]="['active']" directive, everything works as expected.

As a workaround I'm currently using [hidden] instead of the *ngIf, but of course I would prefer to remove the DOM while not displayed.

**Expected behavior**

ngOnInit of all contained components only running once per ngIf expression change to true and ngOnDestroy running once on expression changed to false.

**Please tell us about your environment:**

Using Angular 2, Angular-Meteor, WebStorm, NPM

* **Angular version:** 2.4.5

* **Browser:** 

Chrome 55.0.2883.87
 
* **Language:**
Typescript 2.1.5
ES5

**Comment by DzmitryShylovich** [comment]
Pls add a plunkr with repro

**Comment by stefanholzapfel** [comment]
Some more things I found out:
When I downgrade the router to version 3.3.0 everything works well. With version 3.4.1 the effect is back.

I have also substituted my meteor-angular observable by a setTimeout function to avoid zone/change detection issues coming from that side (error remains).

I will try to build a plunkr, but I have troubles getting the angular router running. Maybe anyone has a running plunkr with angular router version 3.4.1 or higher to share? Thx!

**Comment by DzmitryShylovich** [comment]
http://plnkr.co/edit/RGnYoupY8jAB4MGyWbN0?p=preview

**Comment by stefanholzapfel** [comment]
Thanks!

I have built a primitive example (without ngFor) but you can already see, that the ngOnInit runs two times:

http://plnkr.co/edit/7cFyhKFc1xDA8hyvACHj?p=preview

**Comment by DzmitryShylovich** [comment]
yeah, it's `routerLinkActive` bug

**Comment by stefanholzapfel** [comment]
Is it a known bug? Didn't find anything about it... Thx

**Comment by argasek** [comment]
@stefanholzapfel If you mean by "known" is "did anyone else experience it", then yes, we've just hit this problem when upgrading to from Angular 2.0 to 2.4 recently. I also created a repo with reproduction of the issue (https://github.com/argasek/angular2-nested-ngoninit-issue) and was about to create a bug report, but I found yours and I no longer need to open a new one. 👍

**Comment by stefanholzapfel** [comment]
I was just confused because I have read Dzmitry's post like "yeah, it's THE routerLinkActive bug". But I guess she meant "yeah, it's A routerLinkActive" bug.

Nevermind, as long as it gets fixed I'm happy :)

**Comment by Saegrov** [comment]
We just ran into this bug using `routerLinkActive` in combination with our own custom structural directive.
We downgraded to Angular 2.4.3, which solved our issue.

**Comment by nipuna777** [comment]
We are facing an issue similar to what @Saegrov mentioned. 

When using custom components inside an *ngFor, 
[plnkr link here](http://plnkr.co/edit/qmgIQ2vxzaZkhg34IKVw?open=app%2Fapp.component.ts&p=preview)

The highlighting does not work when `second in loop` is clicked. This is a custom `menu-item` component that is being generated from a `*ngFor`. 

Should I open a new issue regarding this? Will this be fixed with this issue or am I doing something wrong?

**Comment by ssarj** [comment]
Just mentioning for completeness; but bumped into the same (I'm guessing) issue with `ngOnInit` called multiple times, though without specific usage of `ngFor`, or `ngIf`. [Plnkr](https://plnkr.co/edit/uD6wqnvPfz8rHuQzNpdk?p=preview)

**Comment by rerickson1** [comment]
I am seeing ngOnInit called multiple times on the same component on Angular 4.0.0 as well.  I haven't been able to nail down what's causing it. 
I previously downgraded to Angular 2.4.3 to make the issue go away.  I am now planning to change my component to hook into ngAfterViewInit() for the time being...

**Comment by kareljuricka** [comment]
I'm facing the same problem as @ssarj mentioned. Using Angular 4.0.2.

**Comment by DenisVuyka** [comment]
We had to pin to 3.x router because of this issue, no way upgrading to 4.x. Are there any plans to fix this problem? It has been open since Feb 1 and several people confirmed it already.

**Comment by santoshkt** [comment]
I have similar problem using ng 2.4.7 and router 3.4.7. I am using my custom structural directive with element that has routerLinkActive. I don't see ngOnInit getting called twice. I downgraded router to 3.3.0 and that cause some other strange issue. I din't try downgrading ng. I ended up adding the custom director on a ng-container which I don't like but I spent 5 hrs before arriving on this page this already.. argh.. 

```
<li routerLinkActive="active">
  <ng-container *wpHasAnyPermission="[newAuthority(permEnum.USER_CREATE, levelEnum.ACCOUNT_OWN)]">
    <a id="createUser" routerLink="users/create">
      <iclass="icon-plus-circle2"></i>Create a New User
    </a>
  </ng-container>
</li>
```

**Comment by vilic** [comment]
@DzmitryShylovich Any progress? 4.4.3 here.

**Comment by atscott** [comment]
Hi there, is someone able to confirm that this is still an issue in v9? I tried to reproduce it but was unable to.
https://stackblitz.com/edit/angular-ivy-cyupae?file=src%2Fapp%2Fapp.module.ts

**Comment by pkozlowski-opensource** [comment]
I'm going to close this one for now given that we don't have any reproduce scenario with the recent Angular. Happy to re-open if / when we get a reproduce scenario.
